libdrgn/iterators.c

```diff
@@ -0,0 +1,461 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "iterators.h"
+#include "helpers.h"
+#include "type.h"
+
+#define RADIX_TREE_ENTRY_MASK ((uint64_t)3)
+
+DEFINE_VECTOR_FUNCTIONS(frames)
+
+static struct drgn_error *
+radix_tree_iter_run_frame(struct radix_tree_iter *iter, bool *is_exhausted);
+
+static struct drgn_error *
+radix_tree_iter_handle_node(struct radix_tree_iter *iter,
+			    struct drgn_object *node, uint64_t index)
+{
+	struct drgn_error *err;
+	struct radix_tree_iter_frame *frame = NULL;
+	uint64_t value;
+
+	err = drgn_object_read(node, node);
+	if (err)
+		goto out;
+	err = drgn_object_read_unsigned(node, &value);
+	if (err)
+		goto out;
+	if (((value & RADIX_TREE_ENTRY_MASK) !=
+	     iter->RADIX_TREE_INTERNAL_NODE)) {
+		// Base-case, node is NOT internal
+		iter->entry.node = *node;
+		iter->entry.index = index;
+		return NULL;
+	}
+
+	// We are dealing with an internal node, and must iterate over its slots
+
+	err = drgn_object_set_unsigned(node, iter->node_type,
+				       value & ~iter->RADIX_TREE_INTERNAL_NODE,
+				       0);
+	if (err)
+		goto out;
+	frame = frames_append_entry(&iter->frames);
+	if (!frame) {
+		err = &drgn_enomem;
+		goto out;
+	}
+	frame->index = index;
+	frame->loop_index = 0;
+	struct drgn_object tmp;
+	drgn_object_init(&tmp, iter->prog);
+	drgn_object_init(&frame->slots, iter->prog);
+	err = drgn_object_member_dereference(&tmp, node, "shift");
+	if (err)
+		goto out;
+	err = drgn_object_read_unsigned(&tmp, &frame->shift);
+	if (err)
+		goto out;
+	err = drgn_object_member_dereference(&frame->slots, node, "slots");
+	if (err)
+		goto out;
+
+out:
+	if (err) {
+		if (frame) {
+			drgn_object_deinit(&frame->slots);
+			frames_pop(&iter->frames);
+		}
+		return err;
+	}
+
+	bool _;
+	return radix_tree_iter_run_frame(iter, &_);
+}
+
+static struct drgn_error *
+radix_tree_iter_run_frame(struct radix_tree_iter *iter, bool *is_exhausted)
+{
+	while (iter->frames.size) {
+		struct radix_tree_iter_frame *frame =
+			&iter->frames.data[iter->frames.size - 1];
+		if (frame->loop_index <= iter->RADIX_TREE_MAP_MASK) {
+			struct drgn_object node;
+			drgn_object_init(&node, iter->prog);
+			struct drgn_error *err = drgn_object_subscript(
+				&node, &frame->slots, frame->loop_index);
+			if (err)
+				return err;
+			*is_exhausted = false;
+			return radix_tree_iter_handle_node(
+				iter, &node,
+				frame->index +
+					(frame->loop_index++ << frame->shift));
+		} else {
+			drgn_object_deinit(&frame->slots);
+			frames_pop(&iter->frames);
+		}
+	}
+	*is_exhausted = true;
+	return NULL;
+}
+
+struct drgn_error *radix_tree_iter_init(struct drgn_program *prog,
```

Comment:
Most of this function is directly out of `linux_helper_radix_tree_lookup`, right? `RADIX_TREE_ENTRY_MASK` is also defined there. I think it'd be best to move the iterator implementations to `linux_kernel_helpers.c` so that you can share the common code in static functions.

Reply:
I'll make this change  _very last_ after doing everything else, because presently I've been rebasing in all my changes in response to your suggestions so I can continue to test them against the old pure-Python implementations of the linux helpers (which are removed by the commit @HEAD on this branch). Moving all of these functions to `linux_kernel_helpers.c` will almost certainly make it impossible to continue with this workflow,  as attempting to move any changes to _before_ the most recent commit on this branch at that point would generate merge conflicts.

Reply:
This has been split off into a separate PR: #133